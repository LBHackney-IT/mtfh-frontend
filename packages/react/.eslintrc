{
  "extends": "@lbh-mtfh/eslint-config",
  "parserOptions": {
    "project": "tsconfig.json"
  },
  "rules": {
    "prefer-arrow-callback": "off",
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": ["**/*/rollup.config.js"]
      }
    ],
    "jsx-a11y/anchor-is-valid": [
      "error",
      {
        "components": ["Link"],
        "aspects": ["invalidHref"]
      }
    ]
  }
}
