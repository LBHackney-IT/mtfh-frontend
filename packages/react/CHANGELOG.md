# @lbh-mtfh/react

## 1.1.0

### Minor Changes

- 37b7b3e: Initial release

### Patch Changes

- Updated dependencies [37b7b3e]
  - @lbh-mtfh/system@1.1.0
  - @lbh-mtfh/test-utils@1.1.0
