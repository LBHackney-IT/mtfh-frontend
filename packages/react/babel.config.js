const babel = require("../../babel.config")

module.exports = babel
