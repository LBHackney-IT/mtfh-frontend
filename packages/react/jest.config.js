const baseConfig = require("../../jest.config")

baseConfig.setupFilesAfterEnv.push("<rootDir>/jest.setup.ts")

module.exports = {
  ...baseConfig,
}
