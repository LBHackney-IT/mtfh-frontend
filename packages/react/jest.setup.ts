import * as LBHFrontend from "lbh-frontend"

window.LBHFrontend = LBHFrontend
