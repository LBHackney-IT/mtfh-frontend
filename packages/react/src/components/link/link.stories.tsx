import React from "react";

import { Text } from "../text";
import { Link } from "./link";

export default {
  title: "Link",
  component: Link,
};

export const Base = () => {
  return <Link href="/">Home</Link>;
};

export const DangerLink = () => {
  return (
    <Link href="/delete" variant="danger">
      Delete Record
    </Link>
  );
};

export const MutedLink = () => {
  return (
    <Link href="/activities" variant="muted">
      Past activities
    </Link>
  );
};

export const TextColorLink = () => {
  return (
    <Text>
      Read more about{" "}
      <Link href="/mark" variant="text-colour">
        Mark
      </Link>{" "}
      in our new{" "}
      <Link href="/post/1" variant="text-colour">
        blog post
      </Link>
      .
    </Text>
  );
};

export const NoUnderlineLink = () => {
  return (
    <Link href="/" variant="no-underline">
      Home
    </Link>
  );
};

export const BackLink = () => {
  return (
    <Link href="/back" variant="back-link">
      Home
    </Link>
  );
};

export const ExternalLink = () => {
  return (
    <Link href="http://google.com" isExternal>
      Google
    </Link>
  );
};
