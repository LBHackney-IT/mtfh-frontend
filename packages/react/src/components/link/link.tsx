import React, { forwardRef } from "react";
import type * as Polymorphic from "@radix-ui/react-polymorphic";
import cn from "classnames";

import { widthOverrides } from "../../utils";
import styles from "./styles.module.scss";

export interface LinkProps {
  variant?:
    | "link"
    | "danger"
    | "text-colour"
    | "muted"
    | "back-link"
    | "no-underline"
    | "native";
  isExternal?: boolean;
  override?: number;
}

export type LinkComponent = Polymorphic.ForwardRefComponent<"a", LinkProps>;

export const Link: LinkComponent = forwardRef(function Link(
  {
    as: LinkComp = "a",
    variant = "link",
    isExternal = false,
    className,
    rel,
    target,
    override,
    ...props
  },
  ref,
) {
  const linkClasses = cn(
    variant !== "native" && {
      [`${styles.govukLink} ${styles.lbhLink}`]: variant !== "back-link",
      [`${styles.govukBackLink} ${styles.lbhBackLink}`]: variant === "back-link",
      [styles[`lbh-link--${variant}`]]: variant !== "link" && variant !== "back-link",
      [styles.lbhLinkNoVisitedState]:
        !isExternal && !["danger", "muted", "text-colour"].includes(variant),
    },
    widthOverrides(override),
    className,
  );

  return (
    <LinkComp
      ref={ref}
      className={linkClasses}
      rel={isExternal ? "noopener noreferrer" : rel}
      target={isExternal ? "_blank" : target}
      {...props}
    />
  );
});
