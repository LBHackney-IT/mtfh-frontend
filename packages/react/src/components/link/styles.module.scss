@import "~lbh-frontend/lbh/base";
@import "~govuk-frontend/govuk/core/links";
@import "~lbh-frontend/lbh/core/links";
@import "~lbh-frontend/lbh/components/lbh-back-link/back-link";

.lbh-link--no-underline {
  composes: govuk-link--no-underline;
}
