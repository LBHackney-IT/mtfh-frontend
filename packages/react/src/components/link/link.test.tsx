import React, { useEffect } from "react";
import { render } from "@testing-library/react";
import { testA11y } from "@lbh-mtfh/test-utils";

import { Link } from "./link";

test("it renders correctly", async () => {
  const { container } = render(
    <Link href="http://localhost">Test Button</Link>
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders a variant", () => {
  const { container } = render(
    <Link href="http://localhost" variant="muted">
      Test Button
    </Link>
  );
  expect(container).toMatchSnapshot();
});

test("it renders a back link", () => {
  const { container } = render(
    <Link href="http://localhost" variant="back-link">
      Test Button
    </Link>
  );
  expect(container).toMatchSnapshot();
});

test("it sets the correct attributes on isExternal", async () => {
  const { container } = render(
    <Link href="http://localhost" isExternal>
      Test Button
    </Link>
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});
