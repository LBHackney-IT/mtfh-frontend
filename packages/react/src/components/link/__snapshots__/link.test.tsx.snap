// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a back link 1`] = `
<div>
  <a
    class="govukBackLink lbhBackLink lbhLinkNoVisitedState"
    href="http://localhost"
  >
    Test Button
  </a>
</div>
`;

exports[`it renders a variant 1`] = `
<div>
  <a
    class="govukLink lbhLink lbh-link--muted"
    href="http://localhost"
  >
    Test Button
  </a>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <a
    class="govukLink lbhLink lbhLinkNoVisitedState"
    href="http://localhost"
  >
    Test Button
  </a>
</div>
`;

exports[`it sets the correct attributes on isExternal 1`] = `
<div>
  <a
    class="govukLink lbhLink"
    href="http://localhost"
    rel="noopener noreferrer"
    target="_blank"
  >
    Test Button
  </a>
</div>
`;
